### DISCUSSION TITLE
File input handling

### DISCUSSION THREAD

**Discussion by ksmoore17** [idom-team/idom#373] [Question] [answered]
Hi, this package is so cool.

I'm trying to use an input to get an image from the client into bytes (or anything I can open with PIL on the server side). I've tried to use the `Input` widget with `type='file'`, and the resulting argument to the callback is a fakepath. I can also see that `value` is the only key in the event dictionary with the same fakepath when I create a handler manually and use `idol.html.input`. From my limited js knowledge, the fakepath is not useful and a bytes array from `event['target']['result']` is what I was hoping for.

```python
@events.on("change")
def on_change(event: Dict[str, Any]) -> None:
    value = event['target']['result']

    update_image_bytes(value)

return idom.html.input(
    {
        'id': 'image-input',
        'type': 'file',
        'hidden': True,
        'accept': 'image/*'
    },
    event_handlers=events
)
```

I also think that I could do it if I could access the input like `document.getElementById("#image-input").files[0]`, but I don't know anything about that in idom. I guess that another possibility would be to use a js component here.

To be clear, I don't really want to post/upload, I'm basically just trying to display the file after being b64 encoded and used as `src` in an `image` element (I have been able to get this encoding and displaying working in js). Any help would be greatly appreciated.

**Comment by rmorshea** [comment, marked as answer]
I've created an issue to track this: https://github.com/idom-team/idom/issues/375

**Reply by ksmoore17** [reply]
Hi again. I'm grateful for your quick action, and I'm still enjoying working with idom. I'm trying the 0.28.0 release now, and the `event.data` dictionary in the `Layout.dispatch` call still only contains the "value" key (with fakepath as corresponding value). In other words, I'm not sure if I'm missing something to access the file handle, but I can't see anything new in the python callstack when using an input with `type='file'`.

```python
@idom.component
def ImageInput(set_image_bytes_callback):
    events = idom.Events()

    @events.on("change")
    def on_change(event_data: Dict[str, Any]) -> None:
        value = event_data['files']

        set_image_bytes_callback(value)

    return idom.html.input(
        {
            'type': 'file',
            'hidden': True,
            'accept': 'image/*'
        },
        event_handlers=events
    )
```

This is basically what I'm working with now, where `event_data` has only a `"value"` key

**Reply by rmorshea** [reply]
I'll take a look at this. I only wrote a client-side test for this so something must be wrong when running end-to-end.

**Reply by rmorshea** [reply]
I'm on `main` so maybe there's a difference, but I'm actually getting a `files` key, it's just not serializing correctly. The following code:

```python
import idom

@idom.component
def Test():
    return idom.html.input({"type": "file", "id": "my-input", "onChange": print})

idom.run(Test, port=8000)
```

Produces the following output when a file is added:

```
{'value': 'C:\\fakepath\\some-file.txt', 'files': {'0': {}}}
```

**Reply by rmorshea** [reply]
@ksmoore17 I think the reason you're not seeing the latest updates is due to a bug in how the JS build dir is updated. You should get the same output as me if you run `idom restore`. This will be fixed in the next release.

**Reply by ksmoore17** [reply]
Seeing that files key now! And with a dict representation of the file object. Realizing now that I will need to use the js [FileReader](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) api or something similar (`File.arrayBuffer()`) since the python side can't really work with the files from the "handles" (the dicts returned in the `files` key). I tried for a bit to return it in those dicts in the hasFiles callback (tacking it on as a string). This was the general idea:
```js
files: Array.from(target.files).map((file) => {
  var bytes;
  file.arrayBuffer().then((buffer) => {bytes = buffer});

  return {
    lastModified: file.lastModified,
    name: file.name,
    size: file.size,
    type: file.type,
    bytes: bytes
  }
})
```

This seems like a dead end because it's an async process, and it was supposed to be a hack anyway. I think it would be nice to have a python `idom.util` function or something like a `useFile` hook that can take this file handle that is being successfully sent to the python event handler and access the `FileReader` api on the browser to send back the bytes array. Let me know what you think, I would be interested in working on this.

**Reply by rmorshea** [reply]
I think this might be best written as [a custom component](http://idom-docs.herokuapp.com/docs/javascript-components.html#custom-javascript-components), the API for which has been subject to change in the past. However I think  it'll be mostly finalized in the next release so once it's out I can make the necessary changes and remove the warning from the [template repository](https://github.com/idom-team/idom-react-component-cookiecutter).

**Reply by rmorshea** [reply]
I can imagine the JS module might look like this:

```javascript
export function createElement() {};
export function renderElement() {};
export function unmountElement() {};

export function FileInput({ onChange }) {
  const handleChange = (event) => {
    const fileDataPromises = Array.from(event.target.files).map(promiseFileData);
    Promise.all(fileDataPromises).then((files) => onChange({ files: files });
  };
  return <input onChange={handleChange} />
} 

function promiseFileData(file) {
  return file.arrayBuffer().then((bytes) => (
    {
      lastModified: file.lastModified,
      name: file.name,
      size: file.size,
      type: file.type,
      bytes: bytes
    }
  ))
}
```
